core/amber/src/main/python/core/runnables/data_processor.py

```diff
@@ -0,0 +1,186 @@
+from proto.edu.uci.ics.amber.engine.common import ActorVirtualIdentity, ControlInvocationV2, ControlPayloadV2, \
+    LinkIdentity, ReturnInvocationV2
+
+
+class DataProcessor(StoppableQueueBlockingRunnable):
+
+    def __init__(self, input_queue: InternalQueue, output_queue: InternalQueue, udf_operator: UDFOperator):
+        super().__init__(self.__class__.__name__, queue=input_queue)
+
+        self._input_queue: InternalQueue = input_queue
+        self._output_queue: InternalQueue = output_queue
+        self._udf_operator: UDFOperator = udf_operator
+        self._current_input_tuple: Optional[Union[Tuple, InputExhausted]] = None
+        self._current_input_link: Optional[LinkIdentity] = None
+
+        self.context = Context(self)
+        self._async_rpc_server = AsyncRPCServer(output_queue, context=self.context)
+        self._async_rpc_client = AsyncRPCClient(output_queue, context=self.context)
+
+    @overrides
+    def pre_start(self) -> None:
+        self.context.state_manager.assert_state(WorkerState.UNINITIALIZED)
+        self.context.state_manager.transit_to(WorkerState.READY)
+
+    @overrides
+    def receive(self, next_entry: IQueue.QueueElement) -> None:
+        """
+        Main entry point of the DataProcessor. Upon receipt of an next_entry, process it respectfully.
+
+        :param next_entry: An entry from input_queue, could be one of the followings:
+                    1. a ControlElement;
+                    2. a DataElement.
+        """
+        match(
+            next_entry,
+            DataElement, self._process_data_element,
+            ControlElement, self._process_control_element
+        )
+
+    def process_control_payload(self, tag: ActorVirtualIdentity, payload: ControlPayloadV2) -> None:
+        """
+        Process the given ControlPayload with the tag.
+        :param tag: ActorVirtualIdentity, the sender.
+        :param payload: ControlPayloadV2 to be handled.
+        """
+        # logger.debug(f"processing one CONTROL: {payload} from {tag}")
+        match(
+            (tag, get_one_of(payload)),
+            typing.Tuple[ActorVirtualIdentity, ControlInvocationV2], self._async_rpc_server.receive,
+            typing.Tuple[ActorVirtualIdentity, ReturnInvocationV2], self._async_rpc_client.receive
+        )
+
+    def process_input_tuple(self) -> None:
+        """
+        Process the current input tuple with the current input link. Send all result Tuples
+        to downstream operators.
+
+        This is being invoked for each Tuple/Marker that are unpacked from the DataElement.
+        """
+        if isinstance(self._current_input_tuple, Tuple):
+            self.context.statistics_manager.increase_input_tuple_count()
+
+        for tuple_ in self.process_tuple(self._current_input_tuple, self._current_input_link):
+            self.context.statistics_manager.increase_output_tuple_count()
```

Comment:
do we need to filter out `None` values

Reply:
I changed the API to be Iterator[Optional[Tuple]] and added filtering logic for None yield outputs, also added the check control logic in between `next()` invocation.